### DISCUSSION TITLE
Feature request: Convert variables from mutable to immutable at runtime

### DISCUSSION THREAD

**Discussion by LiamSwayne** [modular/modular#1791] [Ideas]
It would be helpful to be able to make a variable immutable based on control flow or keywords.

If `fix` was the keyword used to convert mutable data types to an immutable state, it could look like this:
```
var x:Int = 1
if runtime_expr:
    fix x

x += 1 # this line will only cause an error if runtime_expr evaluates to True
```

This behavior can be produced with `struct`s, but to my knowledge requires running control flow and checking for mutability after every function call to an instance. I assume there is a much more efficient way to implement a feature like this, even in a `struct`. Perhaps a boolean tacked on to the instance could hold it's value until it is freed by the `fix` call, and subsequent checks for mutability would throw an error when trying to mutate the variable due to being unable to access it's correlating boolean. I am sure there is a better approach, but even a decorator on a `struct` to provide this functionality with some overhead would be very useful.

**Comment by Brian-M-J** [comment]
>I assume there is a much more efficient way to implement a feature like this

>I am sure there is a better approach

`let` proceeds to quietly cry in the corner...

Also, I'm not really sure what you're trying to accomplish here, that can't already be done with Mojo:
```python
var x: Int = 1
if not runtime_expr:
    x += 1 # No runtime mutability checking needed
```

Without a compelling real world example, I don't see a need for something like this.
